ALGORITHMS.md

```diff
@@ -164,48 +188,58 @@ If [OQS_KEM_ENCODERS](CONFIGURE.md#OQS_KEM_ENCODERS) is enabled the following li
 
 |Algorithm name |    default OID    | environment variable |
 |---------------|:-----------------:|----------------------|
-| frodo640aes | 1.3.9999.99.50 | OQS_OID_FRODO640AES
-| p256_frodo640aes | 1.3.9999.99.49 | OQS_OID_P256_FRODO640AES
-| x25519_frodo640aes | 1.3.9999.99.38 | OQS_OID_X25519_FRODO640AES
-| frodo640shake | 1.3.9999.99.52 | OQS_OID_FRODO640SHAKE
-| p256_frodo640shake | 1.3.9999.99.51 | OQS_OID_P256_FRODO640SHAKE
-| x25519_frodo640shake | 1.3.9999.99.39 | OQS_OID_X25519_FRODO640SHAKE
-| frodo976aes | 1.3.9999.99.54 | OQS_OID_FRODO976AES
-| p384_frodo976aes | 1.3.9999.99.53 | OQS_OID_P384_FRODO976AES
-| x448_frodo976aes | 1.3.9999.99.40 | OQS_OID_X448_FRODO976AES
-| frodo976shake | 1.3.9999.99.56 | OQS_OID_FRODO976SHAKE
-| p384_frodo976shake | 1.3.9999.99.55 | OQS_OID_P384_FRODO976SHAKE
-| x448_frodo976shake | 1.3.9999.99.41 | OQS_OID_X448_FRODO976SHAKE
-| frodo1344aes | 1.3.9999.99.58 | OQS_OID_FRODO1344AES
-| p521_frodo1344aes | 1.3.9999.99.57 | OQS_OID_P521_FRODO1344AES
-| frodo1344shake | 1.3.9999.99.60 | OQS_OID_FRODO1344SHAKE
-| p521_frodo1344shake | 1.3.9999.99.59 | OQS_OID_P521_FRODO1344SHAKE
+| frodo640aes | 1.3.9999.99.58 | OQS_OID_FRODO640AES
```

Comment:
Looks like all KEM OIDs changed, is this necessary?

Reply:
These are generated OIDs, i.e. ones that no-one ever cared about (to specify). I'm thus inclined to leave as-is (and take complaints as indications that they ought to be defined in "generate.yml" -- with the complaints pointing to a spec about whose existence I'd gladly learn).

Reply:
Fine with me to use generated ones with the absence of specified ones. Just wondering why the PR updates the OIDs of algorithms unrelated to ML-KEM/DSA (e.g., frodo640aes:`1.3.9999.99.50` -> `1.3.9999.99.58`), did these algorithms change?

Reply:
> why the PR updates the OIDs of algorithms

https://github.com/open-quantum-safe/oqs-provider/blob/510fea62449edb14b3ead8a51061b251817298cf/oqs-template/generate.py#L121-L125

Reply:
Thanks for the pointer, I see two issues with the get_tmp_kem_oid() code.

https://github.com/open-quantum-safe/oqs-provider/blob/510fea62449edb14b3ead8a51061b251817298cf/oqs-template/generate.py#L95-L98

- OIDs of unrelated algorithms change if new ones are added: e.g., frodo640aes: `1.3.9999.99.50` -> `1.3.9999.99.58`
- OID collisions: e.g., before the PR, OID `1.3.9999.99.54` was for frodo976aes, now the same OID is for `x448_bikel3`

But maybe I misunderstand the purpose of the temp KEM OIDs. If they are just dummy values not supposed to be globally unique, why not assign the same temp OID for all algorithms without specified OID?

Reply:
> why not assign the same temp OID for all algorithms without specified OID?

Such as to avoid persistent artifacts of different algorithms generated from the same `oqsprovider` version to have the same OID. This has been done to ensure hackathon events (using one `oqsprovider` version) can differentiate artifacts created.

You do have a point in that different `oqsprovider` versions can generate OID clashes "over time": This has not been considered a problem as unspecified OIDs cannot/should not be relied on "over time", though: If someone would consider this an issue, s/he is free to specify OIDs in "generate.yml" to avoid that -- but then also has to "manage" those OIDs.

Reply:
Thanks for the background, I wasn't aware of the KEM temp OIDs for the hackathons.

> different `oqsprovider` versions can generate OID clashes "over time"

Would it be reasonable to add a warning for the KEM default OIDs to [ALGORITHMS.md](https://github.com/open-quantum-safe/oqs-provider/blob/main/ALGORITHMS.md#OIDs)? Please correct me if I'm wrong, but at least for the signatures I think we so far attempted to avoid OID clashes over time also for the "OQS-defined" OIDs. I read the [warning](https://github.com/open-quantum-safe/oqs-provider/blob/main/ALGORITHMS.md#OIDs) that "X.509 OIDs may be subject to change prior to final standardization." such as the default OIDs are non-final, but not that they might clash over time.

Reply:
> Fine with me to use generated ones with the absence of specified ones. Just wondering why the PR updates the OIDs of algorithms unrelated to ML-KEM/DSA (e.g., frodo640aes:`1.3.9999.99.50` -> `1.3.9999.99.58`), did these algorithms change (in liboqs)?

I think we should keep consistent with OIDs we've used in the past.

Reply:
> I think we should keep consistent with OIDs we've used in the past.

This is a plausible goal, but I do not want to resolve this within this PR to not bloat the scope of this PR (introducing ML- algorithms). Complicating matters, this feature has been introduced 2 releases ago, so there is not a single "past" as per the comment above. Further, persistent KEMs are not an active feature by default, so not resolving this now is not considered harmful.

To resolve this discussion, #351 has been created to track the underlying issue and clarification has been added to the documentation in https://github.com/open-quantum-safe/oqs-provider/pull/348/commits/29e9e808128b33e1a873a6c23ca3b20d79020078 to properly set user expectations. Contributions to resolve for good very welcome, preferably before a next release.

Reply:
> > I think we should keep consistent with OIDs we've used in the past.
> 
> This is a plausible goal, but I do not want to resolve this within this PR to not bloat the scope of this PR (introducing ML- algorithms). Complicating matters, this feature has been introduced 2 releases ago, so there is not a single "past" as per the comment above. Further, persistent KEMs are not an active feature by default, so not resolving this now is not considered harmful.
> 
> To resolve this discussion, #351 has been created to track the underlying issue and clarification has been added to the documentation in [29e9e80](https://github.com/open-quantum-safe/oqs-provider/commit/29e9e808128b33e1a873a6c23ca3b20d79020078) to properly set user expectations. Contributions to resolve for good very welcome, preferably before a next release.

If you want to do the dev work for it in a separate PR / branch, okay, but if so then wouldn't that need to be resolved before this PR can land?  Otherwise, the OIDs will change when this PR lands, and then change back (!?) when the second PR lands?  That seems like it will cause even more chaos.

Reply:
Unless I've misunderstood the support for KEMs in OQS Provider.  Where do OIDs for KEMs actually get used?  Not in TLS, since TLS key exchange has its own namespace.  OIDs would come up in certificates and S/MIME -- does OQS Provider support generation of X.509 certs with KEM public keys as the subject public key?

Reply:
> does OQS Provider support generation of X.509 certs with KEM public keys as the subject public key?

No(t that I knew): How would the cert signature be done in such scenario?

Reply:
> the OIDs will change when this PR lands, and then change back (!?) when the second PR lands? That seems like it will cause even more chaos.

The OIDs have changed already. They're random, after all -- since October 2023.

Reply:
> > does OQS Provider support generation of X.509 certs with KEM public keys as the subject public key?
> 
> No(t that I knew): How would the cert signature be done in such scenario?

The subject's public key can be a KEM / public key encryption key, but the CA must use a digital signature scheme key in order to produce a signature on such a certificate.

Reply:
So how is this facilitated using `openssl`? I for sure have never done this. If you know how to, please contribute a test for this.

Reply:
> > the OIDs will change when this PR lands, and then change back (!?) when the second PR lands? That seems like it will cause even more chaos.
> 
> The OIDs have changed already. They're random, after all -- since October 2023.

Did they change in October 2023, or were they introduced in October 2023?  I'm looking at https://github.com/open-quantum-safe/oqs-provider/commit/5515b4b4d084b2f06a7a935b9c6e796ae070de15#diff-ef6f384f213c1efa6cdc7974fec75a672ce414650acb8ec1f0abdd5984076595R167 which is dated October 5 2023 and it looks to me like this when OIDs for e.g. FrodoKEM were first added, and then they've been stable since them.